### ISSUE TITLE
uWSGI reports full listen queue

### ISSUE THREAD

**Issue by emmatyping** [getsentry/self-hosted#1573]
### Version

22.7.0.dev0

### Steps to Reproduce

1. Run `./install.sh`
2. `docker compose up -d`
3. `docker compose logs web`

### Expected Result

No uWSGI errors.

### Actual Result

The web container logs contain

```
uWSGI listen queue of socket "127.0.0.1:46721" (fd: 3) full
```

It seems that uWSGI is reporting that the socket connection queue is full. This is weird for 2 reasons:

 - This is a brand new instance of sentry self-hosted
 - The limit is 100 connections(!)

Edit: there is some more info on the listen queue in the uWSGI docs https://uwsgi-docs.readthedocs.io/en/latest/articles/TheArtOfGracefulReloading.html#the-listen-queue

**Comment by aminvakil** [comment]
We receive 300 requests per second on average on our sentry instance and do not have a problem, probably because we're using a nginx installed on host to reverse proxy to sentry instance, so we're not hitting sentry with 300 requests directly.

I guess it's reasonable for any uWSGI instance to use a reverse proxy in front of it instead of delivering requests to it directly.

Do you think it suffices if we document this?

**Comment by aminvakil** [comment]
@ValentinFrancois Sorry for late response.

Based your configuration `nginx` can keep a limited set of open connections to backend you've configured, and uses those connections to pass the traffic.

Therefore there is no need to open a new connection to backend for each new connection.

**Comment by ValentinFrancois** [comment]
@aminvakil you're mentioning [this mechanism](https://www.nginx.com/blog/http-keepalives-and-web-performance/), right? I think I understand better now
![image](https://user-images.githubusercontent.com/24223132/196034849-55285e5a-6a07-47be-9063-63ee9697adfe.png)

**Comment by aminvakil** [comment]
@ValentinFrancois Exactly!

**Comment by Nekuromento** [comment]
Encountering this error on M1 mac host (I think this is caused by https://github.com/unbit/uwsgi/issues/2406)
Anyone managed to find a workaround?

**Comment by yanghua-ola** [comment]
> We receive 300 requests per second on average on our sentry instance and do not have a problem, probably because we're using a nginx installed on host to reverse proxy to sentry instance, so we're not hitting sentry with 300 requests directly.
> 
> I guess it's reasonable for any uWSGI instance to use a reverse proxy in front of it instead of delivering requests to it directly.
> 
> Do you think it suffices if we document this?

I faced the same issue. If I am not mistaken, the latest docker compose file included an NGINX in front of the web service. Is that any different from running a reverse proxy on the host?

**Comment by aminvakil** [comment]
> > We receive 300 requests per second on average on our sentry instance and do not have a problem, probably because we're using a nginx installed on host to reverse proxy to sentry instance, so we're not hitting sentry with 300 requests directly.
> > I guess it's reasonable for any uWSGI instance to use a reverse proxy in front of it instead of delivering requests to it directly.
> > Do you think it suffices if we document this?
> 
> I faced the same issue. If I am not mistaken, the latest docker compose file included an NGINX in front of the web service. Is that any different from running a reverse proxy on the host?

Yes, it does not keep open connections and just passes them through.

**Comment by aminvakil** [comment]
I've just faced this error myself :)
```
web-1  | 2024-10-08T10:53:12.953684904Z worker 3 lifetime reached, it was running for 86401 second(s)
web-1  | 2024-10-08T10:59:41.334198402Z Tue Oct  8 10:59:41 2024 - *** uWSGI listen queue of socket "127.0.0.1:39041" (fd: 3) full !!! (101/100) ***
```
And flooding `uWSGIlisten queue of socket` afterwards.

I've searched and it seems like this is an issue with uWSGI itself, it should respawn workers after a time or requests (which has already been configured in `sentry/sentry.conf.py`), but seems it does not in certain situations.
https://github.com/unbit/uwsgi/issues/2527#issuecomment-1493527398

cc @guoard

**Comment by guoard** [comment]
In [this comment](https://github.com/unbit/uwsgi/issues/2527#issuecomment-1493527398), we see the message: `Waiting for the GIL`.

Both workers and threads can be used to increase concurrency. Threads are generally more lightweight than workers and consume fewer resources. However, due to Python's [Global Interpreter Lock (GIL)](https://wiki.python.org/moin/GlobalInterpreterLock), threads may not always run in true parallel.

One possible solution to address the deadlock issue could be to increase the number of workers and configure each worker to run with just one thread:

```ini
workers = 6
threads = 1
```

This way, you avoid shared memory between threads and may bypass GIL-related contention.

**Comment by molind** [comment]
@aminvakil Same issue on my side. uWSGI is behind nginx.

**Comment by saksforever** [comment]
+1, restarting web container helps

**Comment by iamironz** [comment]
The same issue with setup, only restart helps

**Comment by andreydruz** [comment]
Temporary fix:

```
* * * * * docker ps -f health=unhealthy --format "docker restart d0374d7a8f8b00d5f473d036eafc5e86115eb74e69d339aaf28c2fba919f712f" | sh
```

**Comment by andreydruz** [comment]
Possible solution:

1. **Increase the Listen Queue Size**: Adjust the `listen` parameter in uWSGI configuration to increase the maximum size of the listen queue. For example:
   ```
   [uwsgi]
   listen = 200
   ```

2. **Increase Number of Workers**: Increase the number of uWSGI worker processes to handle more concurrent requests. For example:
   ```
   [uwsgi]
   processes = 4  # Increase this as necessary
   ```

I which file I can do the change?

**Comment by aldy505** [comment]
I just got this a few hours ago. Restarting the `web` container helps. The easiest solution would be to document that workaround.

**Comment by aminvakil** [comment]
https://uwsgi-docs.readthedocs.io/en/latest/articles/SerializingAccept.html#uwsgi-developers-are-fu-ing-cowards:

```
Linux pthread robust mutexes are solid, we are “pretty” sure about that, so you should be able to enable --thunder-lock on modern Linux systems with a 99.999999% success rates, but we prefer (for now) users consciously enable it
```

**Comment by aldy505** [comment]
@aminvakil do you remember where did we talk about this issue? Was it Discord? Also what was the possible solution back then?

**Comment by aminvakil** [comment]
It was on Discord on self-hosted channel, and I had mentioned my colleague's comment in https://github.com/getsentry/self-hosted/issues/1573#issuecomment-2399715297 there and that we haven't faced this issue since this change.

I'm not sure about changing it for all users though.

**Comment by guoard** [comment]
> I'm not sure about changing it for all users though.

As an alternative, I could suggest using uWSGI’s [harakiri](https://uwsgi-docs.readthedocs.io/en/latest/Options.html#harakiri) option. It automatically kills requests that run too long, which can help avoid stuck workers caused by things like GIL issues or deadlocks.

This might be a safer workaround without needing to change the default setup.

**Comment by aldy505** [comment]
> > I'm not sure about changing it for all users though.
> 
> As an alternative, I could suggest using uWSGI’s [harakiri](https://uwsgi-docs.readthedocs.io/en/latest/Options.html#harakiri) option. It automatically kills requests that run too long, which can help avoid stuck workers caused by things like GIL issues or deadlocks.
> 
> This might be a safer workaround without needing to change the default setup.

@guoard WDYT of making the `harakiri` option as a default one on sentry.conf.py? Is it wise?

**Comment by guoard** [comment]
> [@guoard](https://github.com/guoard) WDYT of making the `harakiri` option as a default one on sentry.conf.py? Is it wise?

Yes, I think it would be a wise default.

There’s already a `proxy_read_timeout` of 30 seconds in the default [`nginx.conf`](https://github.com/getsentry/self-hosted/blob/master/nginx.conf#L67).

If you want to be more cautious, you can set `harakiri = 30` and increase `proxy_read_timeout` to 35s.
Alternatively, you could set `harakiri = 25` to stay within the existing 30s timeout.
